### DISCUSSION TITLE
Proposal for Adding FDS Installation via Spack to the Wiki

### DISCUSSION THREAD

**Discussion by kjrstory** [firemodels/fds#13543] [Ideas]
Hello,

I have written an installation script for FDS using the package manager [Spack](https://github.com/spack/spack). However, it seems that not many people are using it yet. I believe that if I include the following information in this wiki, more people would benefit from it.

I would like to ask in advance if it is okay to add this to the wiki in this way. If it is not desired, I plan to post the information on my personal blog instead.

---------------------------------------------------------------------

To simplify the installation of the source code, FDS can be installed using Spack. The following guide explains how to install FDS using Spack on Ubuntu 20.04.

Step 1. Install Spack

To use Spack, install the following prerequisite packages:

```
apt update
apt install build-essential ca-certificates coreutils curl environment-modules gfortran git gpg lsb-release python3 python3-distutils python3-venv unzip zip
```

Spack can be installed by cloning the GitHub repository without any additional prerequisites:

```
git clone -c feature.manyFiles=true --depth=2 https://github.com/spack/spack.git
```
Once installed, apply the environment variables with the following command (for bash environment):

```
. spack/share/spack/setup-env.sh
```
Now, you can check the installation information for the FDS package using the spack info fds command:

Step 2. Configure the Compiler in Spack
...
Step 3: Install FDS using GNU Fortran (GCC) + Open MPI
...
Step 4: Install FDS using Intel oneAPI + Intel MPI
...
Step 5: Install FDS using Intel oneAPI + Intel MPI + OpenMP
...

**Comment by gforney** [comment]
the information in your post is incomplete.  I have a virtual ubuntu 20.04
system on my pc.  I tried installing spack (and Spack) using apt but apt
could not find it.  I then was able to install it using your git clone
command.  I ran the setup-env.sh (note: the script was not executable).
wasn't clear to me what the the spack info fds command was  Maybe I'm
missing something but I didn't see how to implement steps 2 through 5.  I
suggest you add some more details, and repost and ask other readers of the
forum to try it out if they are interested.  what does spack do that the
fds installation script doesn't do?

On Tue, Oct 8, 2024 at 4:46 AM kjrstory ***@***.***> wrote:

> Hello,
>
> I have written an installation script for FDS using the package manager
> Spack <https://github.com/spack/spack>. However, it seems that not many
> people are using it yet. I believe that if I include the following
> information in this wiki, more people would benefit from it.
>
> I would like to ask in advance if it is okay to add this to the wiki in
> this way. If it is not desired, I plan to post the information on my
> personal blog instead.
> ------------------------------
>
> To simplify the installation of the source code, FDS can be installed
> using Spack. The following guide explains how to install FDS using Spack on
> Ubuntu 20.04.
>
> Step 1. Install Spack
>
> To use Spack, install the following prerequisite packages:
>
> apt update
> apt install build-essential ca-certificates coreutils curl environment-modules gfortran git gpg lsb-release python3 python3-distutils python3-venv unzip zip
>
> Spack can be installed by cloning the GitHub repository without any
> additional prerequisites:
>
> git clone -c feature.manyFiles=true --depth=2 https://github.com/spack/spack.git
> Once installed, apply the environment variables with the following command (for bash environment):
>
> . spack/share/spack/setup-env.sh
> Now, you can check the installation information for the FDS package using the spack info fds command:
>
> Step 2. Configure the Compiler in Spack
> ...
> Step 3: Install FDS using GNU Fortran (GCC) + Open MPI
> ...
> Step 4: Install FDS using Intel oneAPI + Intel MPI
> ...
> Step 5: Install FDS using Intel oneAPI + Intel MPI + OpenMP
> ...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/firemodels/fds/discussions/13543>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6UCRRPQ4PH3AM6XYXMFQTZ2OLXPAVCNFSM6AAAAABPRVNVXGVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGI4DSNJZGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Glenn Forney

**Reply by kjrstory** [reply]
@gforney 

First, I apologize for my previous explanation being insufficient.
I wasn't writing the entire post but rather showing an example of how I intend to structure it.
As for your question, here’s a brief answer:
I am not installing Spack via apt, but rather installing the prerequisite packages with apt. The installation itself is done through git clone.
The setup-env.sh is not an executable file but is used to add Spack-related paths to the PATH environment variable.
There is a period (.) at the beginning of the setup-env.sh line.

I will post the full article in this thread soon.

**Comment by kjrstory** [comment]
## Spack Guide for Installing and Running FDS

This guide was performed on an Ubuntu 20.04 environment with Intel Ice Lake architecture.

### Step 1: Install Spack (Dependencies and GitHub Clone)

To use Spack, install the required dependencies as follows:
```bash
apt update
apt install build-essential ca-certificates coreutils curl environment-modules gfortran git gpg lsb-release python3 python3-distutils python3-venv unzip zip
```

Spack can be installed by cloning the GitHub repository without any additional pre-packages:

```
git clone -c feature.manyFiles=true --depth=2 https://github.com/spack/spack.git
```

Once installed, apply the environment variables as follows. This command is for use in a bash environment:

```
. spack/share/spack/setup-env.sh
```

You can now view the installation information of the FDS package using the spack info fds command:

```
$ spack info fds
MakefilePackage:   fds

Description:
     Fire Dynamics Simulator (FDS) is a large-eddy simulation (LES) code for
    low-speed flows, with an emphasis on smoke and heat transport from
    fires. FDS and Smokeview are free and open-source software tools
    provided by the National Institute of Standards and Technology (NIST) of
    the United States Department of Commerce. Pursuant to Title 17, Section
    105 of the United States Code, this software is not subject to copyright
    protection and is in the public domain. View the full disclaimer for
    NIST-developed software.

Homepage: https://pages.nist.gov/fds-smv

Preferred version:
    6.9.1    [git] https://github.com/firemodels/fds.git at commit 889da6ae08d08dae680f7c0d8de66a3ad1c65375

Safe versions:
    6.9.1    [git] https://github.com/firemodels/fds.git at commit 889da6ae08d08dae680f7c0d8de66a3ad1c65375
    6.9.0    [git] https://github.com/firemodels/fds.git at commit 63395692607884566fdedb5db4b5b4d98d3bcafb
    6.8.0    [git] https://github.com/firemodels/fds.git at commit 886e0096535519b7358a3c4393c91da3caee5072

Deprecated versions:
    None

Variants:
    build_system [makefile]        makefile
        Build systems supported by the package
    openmp [false]                 false, true
        Enable OpenMP support

Build Dependencies:
    gmake  mkl  mpi

Link Dependencies:
    mkl  mpi

Run Dependencies:
    None

Licenses:
    None
```


### Step 2: Configure Spack Compiler

First, configure the compiler for the source code. Use the command below to see the registered compilers:

```
$ spack compilers
==> No compilers available. Run `spack compiler find` to autodetect compilers
```

No compilers have been registered yet. Use the command below to automatically find and register available compilers:

```
$ spack compiler find
==> Added 2 new compilers to /root/.spack/linux/compilers.yaml
    gcc@9.4.0  gcc@8.4.0
==> Compilers are defined in the following files:
    /root/.spack/linux/compilers.yaml
```

On my Ubuntu 20.04, the gcc versions were installed. Let's install intel-oneapi using spack:

```
$ spack install intel-oneapi-compilers
$ spack load intel-oneapi-compilers
$ spack compiler find
==> Warning: The external spec 'intel-oneapi-compilers-classic@=2021.13.1' cannot be used as a compiler: missing the C compiler path under 'extra_attributes:compilers'
==> Added 1 new compiler to /root/.spack/linux/compilers.yaml
    oneapi@2024.2.1
==> Compilers are defined in the following files:
    /root/.spack/linux/compilers.yaml
```

Now we can see that two compilers are registered with Spack:

```
$ spack compilers
==> Available compilers
-- gcc ubuntu20.04-x86_64 ---------------------------------------
gcc@9.4.0  gcc@8.4.0

-- oneapi ubuntu20.04-x86_64 ------------------------------------
oneapi@2024.2.1
```


### Step 3: Install FDS with GNU Fortran (GCC) + Open MPI

To install FDS with the combination of GNU Fortran and Open MPI, use the following command:

```
$ spack install fds@6.9.1 %gcc@9.4.0 ^openmpi
```

This command will install FDS using GNU Fortran (GCC) and Open MPI, and all dependencies will be automatically handled. Below is a portion of the installation log:

```
[+] /shared/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/numactl-2.0.18-wig3qipcmum7g3iwuaxefr45kwa3vbpi
==> Installing pmix-5.0.3-n2n65fd74tsxbhzkmp4epzytdp574zjf [38/44]
==> No binary for pmix-5.0.3-n2n65fd74tsxbhzkmp4epzytdp574zjf found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/3f/3f779434ed59fc3d63e4f77f170605ac3a80cd40b1f324112214b0efbdc34f13.tar.bz2
==> No patches needed for pmix
==> pmix: Executing phase: 'autoreconf'
==> pmix: Executing phase: 'configure'
==> pmix: Executing phase: 'build'
==> pmix: Executing phase: 'install'
==> pmix: Successfully installed pmix-5.0.3-n2n65fd74tsxbhzkmp4epzytdp574zjf
  Stage: 0.72s.  Autoreconf: 0.00s.  Configure: 13.17s.  Build: 47.23s.  Install: 5.88s.  Post-install: 0.93s.  Total: 1m 8.21s
[+] /shared/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/pmix-5.0.3-n2n65fd74tsxbhzkmp4epzytdp574zjf
...
[+] /shared/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/fds-6.9.1-lq5hkyyjvotigvkldh7d3iiigq4xw57r
```

### Step 4: Install FDS with Intel oneAPI + Intel MPI (+ OpenMP)

To install FDS using Intel oneAPI compiler and Intel MPI, use the following command:

```
$ spack install fds %oneapi ^intel-oneapi-mpi
```

To enable OpenMP functionality, use the following command:

```
$ spack install fds %oneapi +openmp ^intel-oneapi-mpi
```

These commands will install FDS with the combination of Intel oneAPI and Intel MPI, with optional OpenMP support. Here, `+` indicates a variant. You can check the available variants using the 'spack info' command. Currently, only the openmp variant exists.

### Step 5: Loading the FDS Environment

Spack allows you to install multiple versions of software simultaneously and switch between them as needed. To check the installed FDS packages, use the following command:

```
$ spack find fds
-- linux-ubuntu20.04-icelake / gcc@9.4.0 ------------------------
fds@6.9.1

-- linux-ubuntu20.04-icelake / oneapi@2024.2.1 ------------------
fds@6.9.1  fds@6.9.1
==> 3 installed packages
```

Let's first use the version of Spack installed with GNU Fortran:

```
$ spack load fds %gcc
$ fds

 Fire Dynamics Simulator

 Current Date     : October  8, 2024  17:46:59
 Revision         : FDS-6.9.1-0-g889da6a-HEAD
 Revision Date    : Sun Apr 7 17:05:06 2024 -0400
 Compiler         : GCC version 9.4.0
 Compilation Date : Oct 08, 2024  17:22:56

 MPI version: 3.1
 MPI library version: Open MPI v5.0.5, package: Open MPI root@ansys-compute-1-01 Distribution, ident: 5.0.5, repo rev: v5.0.5, Jul 22, 2024
 Consult FDS Users Guide Chapter, Running FDS, for further instructions.
```


Now let's use the version of FDS installed with oneAPI:

```
$ spack unload fds
$ spack load fds %oneapi~openmp
$ fds

 Fire Dynamics Simulator

 Current Date     : October  8, 2024  17:47:44
 Revision         : FDS-6.9.1-0-g889da6a-HEAD
 Revision Date    : Sun Apr 7 17:05:06 2024 -0400
 Compiler         : Intel(R) Fortran Compiler for applications running on Intel(R) 64, Version 2024.2.1 Build 20240711
 Compilation Date : Oct 08, 2024 17:36:12

 MPI version: 3.1
 MPI library version: Intel(R) MPI Library 2021.13 for Linux* OS

 Consult FDS Users Guide Chapter, Running FDS, for further instructions.
```

Here, ~ indicates the version without OpenMP. If you use the version with OpenMP, it should work properly:

```
$ spack unload fds 
$ spack load fds %oneapi+openmp
```

If you use the version with OpenMP, it will work as expected.

**Comment by kjrstory** [comment]
Here are the key differences between installing FDS using Spack and the traditional FDS installation script:

1. Dependency Management: With Spack, all required dependencies are automatically handled, whereas the traditional script requires you to manually identify, download, and install each dependency. For FDS, notable dependencies include MPI and MKL, which Spack can install all at once, simplifying the process significantly.

2. Compiler Flexibility: Spack allows you to easily register and manage multiple compilers, such as GCC and Intel oneAPI. The traditional installation often requires you to manually set up and configure the appropriate compiler.

3. Multiple Configurations: Spack supports installing multiple versions of the same software concurrently, enabling easy testing with different compiler and MPI combinations. The traditional installation does not provide this kind of flexibility out of the box.

4. Variants and Features: Spack provides an easy way to include optional features (variants) like OpenMP support with simple commands (+openmp). The traditional method requires modifying the build process manually to add such features.

Spack also uses FDS's Makefile to perform the installation, similar to the traditional method. This means that the core build process remains consistent, but Spack adds enhanced management capabilities on top of it.

@gforney 
Please read the above guide and my opinions and try again.

**Comment by gforney** [comment]
Thanks. Ill take a look. It may be awhile though
